esp-hal/src/uart.rs

```diff
@@ -1633,7 +1656,44 @@ mod asynch {
         }
     }
 
-    impl<T> Uart<'_, T>
+    impl<'d, T> Uart<'d, T, Async>
+    where
+        T: Instance + 'd,
+    {
+        /// Create a new UART instance with defaults in [`Async`] mode.
+        pub fn new_with_config_async<P>(
```

Comment:
I think `new_async_with_config` might be a better choice here, this would give us `new`, `new_with_config`, `new_async`, and `new_async_with_config`, which I think looks reasonable. Happy to hear what others think of course.